### ISSUE TITLE
No schema found / No root module found / Workspace not initialized / Alternative root modules found

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform-ls#424]
### Symptoms

This is a catch-all issue for the following warning messages:

> No schema found for .... Functionality may be limited. You may need to run terraform init.

> No root module found for ... Functionality may be limited. You may need to run terraform init and reload your editor.

> Alternative root modules found for ..., picked: .... You can try setting paths to root modules explicitly in settings.

### Duplicates

This is an issue we'll keep open in place of the following ones:

 - https://github.com/hashicorp/terraform-ls/issues/212
 - https://github.com/hashicorp/terraform-ls/issues/221
 - https://github.com/hashicorp/terraform-ls/issues/207
 - https://github.com/hashicorp/terraform-ls/issues/383

Which doesn't mean the linked issues are in any way invalid, just that it's beneficial to keep the conversation in one place.

All of the above points to one particular root cause, which is that the language server has no schema to use for completion, hover and other functionality.

This may be because one of the following reasons:

 - the directory was not `terraform init`-ed, i.e. plugins were not downloaded, e.g. because
    - a nested module is being edited, which would never be `init`-ed on its own, but only from a parent module
    - terragrunt is in use
 - the directory was `terraform init`-ed, but the LS was unable to obtain the schema, e.g. because
    - remote backend is in use, which makes it difficult to obtain the schema due to authentication workflow
    - tfenv is in use and the version specified in `.terraform-version` is not yet installed
    - terragrunt is in use and therefore plugins cache is decoupled from the relevant directory

The list of reasons is not definitive, but these are the most commonly reported ones.

### Proposed Solution

https://github.com/hashicorp/terraform-ls/issues/354 describes the solution in more detail.

TL;DR of the solution is to obtain the schema through other means (not just via CLI) and combine the different sources, so that there _is_ a schema in most/more cases, then remove both messages from LS entirely.

--- 

Feel free to leave 👍  _reaction_ on this issue, but please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for maintainers and other issue followers and do not help with prioritization.

**Comment by arren-ru** [comment]
I have fix for those guy who uses coc.nvim as me. You need to define rootPatterns parameter with ".terraform"
```vim
let g:coc_user_config.languageserver = #{
            \   terraform: #{
            \       filetypes: ['terraform'],
            \       command: 'terraform-ls',
            \       args: ['serve'],
            \       rootPatterns: ['.terraform'],
            \       initializationOptions: {} },
            \   }
```

**Comment by WhyNotHugo** [comment]
I'm also having this issue, but my scenario is a bit different.

I'm using `terraform-ls` on my local machine, since I want to use its functionality in my code editor (`neovim`). However, running `terraform init` is not possible, since I would need credentials with access to the state and all production resources.

I guess the actual issue comes from `terraform init` doing two things at once: initialising local workspace (which needs to be done by any developer editing terraform code), and loading remote state (something which is far more delicate and generally has much narrower access).

Maybe a flag like `--no-state` would make sense here?

**Comment by radeksimko** [comment]
@WhyNotHugo that scenario would fall under this

> remote backend is in use, which makes it difficult to obtain the schema due to authentication workflow

**Comment by WhyNotHugo** [comment]
Why is this off-topic? Should I then open a separate issue for my use case?

**Comment by radeksimko** [comment]
The scenario you described is what I described above, just in different words I believe?

> Maybe a flag like --no-state would make sense here?

There is an open issue for this request here: https://github.com/hashicorp/terraform/issues/24261
The issue doesn't explicitly mention state - so feel free to mention it there, but the general idea is to have a command/flag which doesn't depend on any factors other than existence of the plugin binary.

**Comment by radeksimko** [comment]
https://github.com/hashicorp/terraform-ls/pull/454 which implements schema cache and relatedly removes the mentioned warnings was just merged.

We plan to cut a release with that change by the end of this week and you are welcomed to try out a snapshot build in the meantime:

 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_amd64.zip](https://github.com/hashicorp/terraform-ls/files/6341718/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_amd64.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_arm64.zip](https://github.com/hashicorp/terraform-ls/files/6341719/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_arm64.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_386.zip](https://github.com/hashicorp/terraform-ls/files/6341720/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_386.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_amd64.zip](https://github.com/hashicorp/terraform-ls/files/6341721/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_amd64.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_arm.zip](https://github.com/hashicorp/terraform-ls/files/6341723/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_arm.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_386.zip](https://github.com/hashicorp/terraform-ls/files/6341724/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_386.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_amd64.zip](https://github.com/hashicorp/terraform-ls/files/6341725/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_amd64.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm.zip](https://github.com/hashicorp/terraform-ls/files/6341726/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm64.zip](https://github.com/hashicorp/terraform-ls/files/6341727/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm64.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_386.zip](https://github.com/hashicorp/terraform-ls/files/6341728/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_386.zip)
 - [terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_amd64.zip](https://github.com/hashicorp/terraform-ls/files/6341729/terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_amd64.zip)


<details>
  <summary>terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_SHA256SUMS</summary>
  
```
22523f65af2bfd5a00ace34d11e0d3992ef8fc1c92e9f8f40fbe66a294e8cfac  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_amd64.zip
2984c952fb5e1ceb5a563e8a20fc7e1a90ae94d000aa379d0c13f3feee13dac2  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_arm.zip
56805bccb228069c250545f4ce7cd90b6fcff07c642ec571019e4a847955485a  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_windows_amd64.zip
57c564bf405d1a6d6f777f5c392c4cb5e3e1b1fb0d333669af7b723020027af7  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_amd64.zip
59ae0097ab2f6abfb57b143e3cbb58e05e169646f8161b51614c48b078694e61  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_darwin_arm64.zip
791311733e163b8f112569fdb3b77f0529f011ad8f3ff026d364dd3e766992cf  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_amd64.zip
85445e27f6e305f2423a659a3846c70a534855c11912bb68835f9dab4019bf63  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_amd64.zip
8f70240541cb1e6e9223fa0b317709b8136774079cc8c9725e85e1a21da6fc51  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_freebsd_386.zip
9c74bd353d0da722628f30ba38ca46a9403225ef0c63309cc606afb3133ff26d  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_openbsd_386.zip
c81b73026fb664564c31f61cb0a41bdfbd798debff88ce1fe7fcfe3126c3b838  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_windows_386.zip
d4dd27d4d106a6ac162f0b9da9883489dc81d3c38721675cb74ba3e78b8a7566  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_solaris_amd64.zip
d67e7dc74660bf89022d145fa1decfe7aa63dbc111beab8ab11864fc4d583819  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm.zip
eeb25c425344610e2937e5fc4d91ae1447ae0ca44fd5f6c1a2439fc0c8640113  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_386.zip
f9efe13288fe98c85a2a7d4d93e7cf7417e990525fd1b8195c07dada6b7554de  terraform-ls_v0.15.0-SNAPSHOT-1de4bb5_linux_arm64.zip
```
</details>

Thank you all for your patience!

**Comment by svanharmelen** [comment]
This one seems to work a lot better, but I notice that I have to save a file twice before it will be formatted. Didn't look into it, but seems reasonably consistent (using coc-nvim) using this snapshot, while I didn't notice that with the latest release.

**Comment by radeksimko** [comment]
@svanharmelen That is an interesting observation, but I don't see how that could relate to the recent changes 🤔 
Specifically formatting is done by executing `terraform fmt -`.

Admittedly there is one query into the module cache, because we used to track TF execution path per module because different modules may require different TF versions from different paths, but feature actually never crystalized into anything, so we may as well just remove that extra query - see https://github.com/hashicorp/terraform-ls/pull/466

What could be going on is that the formatting takes time and the client doesn't wait for response for that long. Sublime Text LSP AFAIK recently introduced 1sec timeout for formatting responses.

However that query alone should not be causing any (significant) delay. What I'd guess might be happening is that execution takes longer for some reason (e.g. [as a result of tfenv](https://github.com/tfutils/tfenv/issues/196))

**Comment by svanharmelen** [comment]
Interesting. Thanks for the insights! I'll try to have a closer look (if time permits) to see if this holds true for my setup as well...

**Comment by radeksimko** [comment]
Just FYI the originally planned release was postponed due to the incident involving [GPG key rotation](https://discuss.hashicorp.com/t/hcsec-2021-12-codecov-security-event-and-hashicorp-gpg-key-exposure/23512), which also means I had some extra time to look into the suspected performance issues and make some improvements in that area:
https://github.com/hashicorp/terraform-ls/pull/472 Thanks again for the nudge!

Hoping to make it part of the next release 🔜 , once the key rotation is finished and VSCode extension had chance to release a version with the new (rotated) key embedded as per https://github.com/hashicorp/vscode-terraform/pull/612

**Comment by svanharmelen** [comment]
Nice @radeksimko! Thanks for the update... Didn't have much time to have a closer look, so looking forward to the release 👍🏻

**Comment by radeksimko** [comment]
Language server 0.16.0 was released yesterday. As mentioned in the [changelog](https://github.com/hashicorp/terraform-ls/blob/v0.16.0/CHANGELOG.md#0160-29-april-2021), this version removes all of the above warnings and should provide schema-dependent functionality in _more_ cases due to the fact that it can now combine different sources.

It also lays down the foundations for future work on https://github.com/hashicorp/terraform-ls/issues/193 and https://github.com/hashicorp/terraform-ls/issues/355 Such additional sources of schema will _then_ improve the coverage (and chances of getting relevant schema) further.

In the meantime you may still experience some inaccurate completion/hover/highlighting data (because newer/older version of provider schema was matched) or lack of the data because it's not available in the preloaded batch of official+partner providers and it was not possible to source it locally, which may have been because

 - the config that was initially opened was invalid https://github.com/hashicorp/terraform-ls/issues/56
 - state is stored remotely and requires authentication https://github.com/hashicorp/terraform-ls/issues/128

Feel free to follow the mentioned issues and comment there if you feel impacted by these and feel free to [file](https://github.com/hashicorp/terraform-ls/issues/new/choose) any new bugs or feature requests you find in the new version(s).

Thanks again for your patience!
